### PR TITLE
Fix parsenum bounds

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
I should have spent more time carefully thinking about the code, instead of trying to use the "suppress warnings" (#254) hammer.

File: util/parsenum.h [hunk 1]
```diff
@@ -121,8 +121,10 @@ _Pragma("clang diagnostic pop")
 			    ((min) <= 0) ? 0 : (min),			\
 			    (uintmax_t)(max), (*(x)), (base),		\
 			    (trailing))),				\
-				((((max) < 0) && (errno == 0)) ?	\
-				    (errno = ERANGE) : 0)),		\
+				(((max) <= INTMAX_MAX) ?		\
```
Comment (cperciva) [inline util/parsenum.h:124]: Not sure what you mean by "if the compiler has to treat (max) as an unsigned value" here.  Does this help if max is an unsigned variable?

Comment (gperciva) [reply util/parsenum.h:124]: Right, it took me a while to decypher the commit message.  The idea is basically sanity-checking the upcoming cast to `intmax_t`.  I mean, if max is `INTMAX_MAX + 1`, then we can't do `((intmax_t)(max) < 0) ?`

... at the time, I thought that I had a reason to split up the change into two commits, but at the moment I'm not recreating my thought process.

**Comment by cperciva** [comment]
I'm still not understanding the "if (max) is higher than INT_MAX" bit.

**Comment by gperciva** [comment]
... right, it looks like the commit message is still bogus.  On my 64-bit system, it looks like anything over `LONG_MAX` triggers the problem.  But that's almost certainly a system-specific thing.

Are we discussing the commit message, or the code change as well?  Here's a simplified version of the problem:
```
$ cat simple.c 
#include <limits.h>
#include <stdint.h>
#include <stdio.h>

#define IS_NEGATIVE(x)					\
	((x) < 0) ? 1 : 0

int main() {
	if (IS_NEGATIVE(SIZE_MAX))
		printf("negative\n");
	else
		printf("positive\n");
}
$ gcc -Wall -Wextra simple.c
simple.c: In function 'main':
simple.c:6:7: warning: comparison of unsigned expression < 0 is always false [-Wtype-limits]
    6 |  ((x) < 0) ? 1 : 0
      |       ^
simple.c:9:6: note: in expansion of macro 'IS_NEGATIVE'
    9 |  if (IS_NEGATIVE(SIZE_MAX))
      |      ^~~~~~~~~~~
$
```

Replacing `SIZE_MAX` with `SIZE_MAX/2` makes it work.  Replacing it with `SIZE_MAX/2 + 1` causes the same problem.

My proposed solution is:

```
$ cat prob.c 
#include <limits.h>
#include <stdint.h>
#include <stdio.h>

#if 0
#define IS_NEGATIVE(x)					\
	((x) < 0) ? 1 : 0
#else
#define IS_NEGATIVE(x)					\
	(((x) <= INTMAX_MAX) ?				\
		(((intmax_t)(x) < 0) ? 1 : 0) :		\
		0)
#endif

int main() {
	if (IS_NEGATIVE(SIZE_MAX))
		printf("negative\n");
	else
		printf("positive\n");

	if (IS_NEGATIVE(UINTMAX_MAX))
		printf("negative\n");
	else
		printf("positive\n");
}
$ gcc -Wall -Wextra prob.c && ./a.out 
positive
positive
td@beastie: ~/tmp
$ 
```

(I should clarify that at this point of `PARSENUM`, we're only dealing with unsigned integers, not floats or signed.)

**Comment by cperciva** [comment]
It looks to me like the warning is being triggered when `max` has an unsigned type.  What you're seeing with constants is probably because literal constants are typed as signed unless the only type they fit into is an unsigned type.

**Comment by gperciva** [comment]
Yes.  But looking at kivaloo's `lib/http/http.c`, would you rather see:
```
if (PARSENUM_EX(&clen, (const char *)buf, 0, SIZE_MAX, 16, 1)) {
```
or
```
if (PARSENUM_EX(&clen, (const char *)buf, 0, (long)SIZE_MAX, 16, 1)) {
```

?

**Comment by cperciva** [comment]
Obviously not `(long)SIZE_MAX` since signed integer overflow is undefined behaviour.

I'm not saying that the code is wrong, just that the commit message is wrong.  The issue isn't how *big* `max` is; the issue is what *type* it has.

**Pull request merged**